libraries/browser-tracker-core/src/tracker/id_cookie.ts

```diff
@@ -0,0 +1,356 @@
+   */
+  sessionIndex: number;
+
+  /**
+   * Index of the current event in the session
+   */
+  eventIndex: number;
+
+  /**
+   * The previous session identifier for this user
+   */
+  previousSessionId: string | null;
+
+  /**
+   * The mechanism that the session information has been stored on the device
+   */
+  storageMechanism: string;
+
+  /**
+   * Identifier of the first event for this session
+   */
+  firstEventId: string | null;
+
+  /**
+   * Date-time timestamp of when the first event in the session was tracked
+   */
+  firstEventTimestamp: string | null;
+}
+
+export function emptyIdCookie() {
+  const idCookie: ParsedIdCookie = ['1', '', 0, 0, 0, '', '', '', '', '', 0];
+  return idCookie;
+}
+
+/**
+ * Parses the cookie values from its string representation.
+ *
+ * @param id Cookie value as string
+ * @param domainUserId Domain user ID to be used in case of empty cookie string
+ * @returns Parsed ID cookie tuple
+ */
+export function parseIdCookie(
+  id: string,
+  domainUserId: string,
+  memorizedSessionId: string,
+  memorizedVisitCount: number
+) {
+  let now = new Date(),
+    nowTs = Math.round(now.getTime() / 1000),
+    tmpContainer;
+
+  if (id) {
+    tmpContainer = id.split('.');
+    // cookies enabled
+    tmpContainer.unshift('0');
+  } else {
+    tmpContainer = [
+      // cookies disabled
+      '1',
+      // Domain user ID
+      domainUserId,
+      // Creation timestamp - seconds since Unix epoch
+      nowTs,
+      // visitCount - 0 = no previous visit
+      memorizedVisitCount,
+      // Current visit timestamp
+      nowTs,
+      // Last visit timestamp - blank meaning no previous visit
+      '',
+      // Session ID
+      memorizedSessionId,
+    ];
+  }
+
+  if (!tmpContainer[sessionIdIndex] || tmpContainer[sessionIdIndex] === 'undefined') {
+    // session id
+    tmpContainer[sessionIdIndex] = uuid();
+  }
+  if (!tmpContainer[previousSessionIdIndex] || tmpContainer[previousSessionIdIndex] === 'undefined') {
+    // previous session id
+    tmpContainer[previousSessionIdIndex] = '';
+  }
+  if (!tmpContainer[firstEventIdIndex] || tmpContainer[firstEventIdIndex] === 'undefined') {
+    // firstEventId - blank meaning no previous event
+    tmpContainer[firstEventIdIndex] = '';
+  }
+  if (!tmpContainer[firstEventTsInMsIndex] || tmpContainer[firstEventTsInMsIndex] === 'undefined') {
+    // firstEventTs - blank meaning no previous event
+    tmpContainer[firstEventTsInMsIndex] = '';
+  }
+  if (!tmpContainer[eventIndexIndex] || tmpContainer[eventIndexIndex] === 'undefined') {
+    // eventIndex – 0 = no previous event
+    tmpContainer[eventIndexIndex] = 0;
+  }
+
+  const parsed: ParsedIdCookie = [
+    tmpContainer[0] as string,
+    tmpContainer[1] as string,
+    parseInt(tmpContainer[2] as string),
+    parseInt(tmpContainer[3] as string),
+    parseInt(tmpContainer[4] as string),
+    tmpContainer[5] ? parseInt(tmpContainer[5] as string) : (tmpContainer[5] as string),
```

Comment:
The problem here is that the two timestamps are optional – there can either be an integer value with the time or it can be blank (empty string when serialized to cookie). I didn't want to default to nowTs or 0 because that would change blank behaviour. I cleaned this up a bit and changed the empty string to undefined, hope that's better.